pyccel/codegen/printing/fcode.py

```diff
@@ -306,133 +298,60 @@ def _print_Module(self, expr):
             body = '{0}\n{1}\n'.format(body, c_funcs)
         # ...
 
+        contains = 'contains\n' if (expr.funcs or expr.classes or expr.interfaces) else ''
 
-        if expr.funcs or expr.classes or expr.interfaces:
-            body = '\n contains\n{0}'.format(body)
+        parts = ['module {}\n'.format(name),
+                 imports,
+                 'implicit none\n',
+                 private,
+                 decs,
+                 interfaces,
+                 contains,
+                 body,
+                 'end module {}\n'.format(name)]
 
-        return ('module {name}\n'
-                '{imports}\n'
-                'implicit none\n'
-                '{private}\n'
-                '{decs}\n'
-                '{interfaces}\n'
-                '{body}\n'
-                'end module\n').format(name=name,
-                                       imports=imports,
-                                       decs=decs,
-                                       private=private,
-                                       interfaces=interfaces,
-                                       body=body)
+        return '\n'.join([a for a in parts if a])
 
     def _print_Program(self, expr):
         self._handle_fortran_specific_a_prioris(self.parser.get_variables(self._namespace))
 
-        name = 'prog_{0}'.format(self._print(expr.name))
-        name = name.replace('.', '_')
-        modules = ''
-        mpi = False
-        #we use this to detect of we are using so that we can add
-        # mpi_init and mpi_finalize in the code instruction
-        # TODO should we find a better way to do this?
-        imports = list(expr.imports)
-        for i in expr.imports:
-            if 'mpi4py' == str(getattr(i.source,'name',i.source)):
-                mpi = True
-
-        imports = '\n'.join(self._print(i) for i in imports)
-        funcs   = ''
+        name    = 'prog_{0}'.format(self._print(expr.name)).replace('.', '_')
+        imports = ''.join(self._print(i) for i in expr.imports)
+        decs    = ''.join(self._print(i) for i in expr.declarations)
         body    = self._print(expr.body)
 
-        # ... TODO add other elements
-        private_funcs = [f.name for f in expr.funcs if f.is_private]
-        private = private_funcs
-        if private:
-            private = ','.join(self._print(i) for i in private)
-            private = 'private :: {}'.format(private)
-        else:
-            private = ''
-        # ...
-
-        decs    = expr.declarations
         vars_to_print = self.parser.get_variables(self._namespace)
         for v in vars_to_print:
             if v not in expr.variables:
                 decs.append(Declare(v.dtype,v))
```

Comment:
This is where the declaration of the additional variables is carried out. If we keep the current order then the last line needs changing to call `_print`

Reply:
Why can't we get all variable declarations straight from the namespace, then?
```
variables = self.parser.get_variables(self._namespace)
decs = ''.join(_print_Declare(Declare(v.dtype, v) for v in variables))
```

Reply:
We can, This is what `Program.declarations` does anyway:
```
    @property
    def declarations(self):
        return [Declare(i.dtype, i) for i in self.variables]
```
I think this would be a better way of doing it as currently we collect the variables to create declarations twice. We need to make sure there is a clear comment about it though so we remember why the ordering is the way it is. The current implementation is clunky but it has the advantage of making us think about it and therefore realise that there are new variables that aren't saved in the Program